### ISSUE TITLE
type of torch.bernoulli and torch.multinomial inconsistent

### ISSUE THREAD

**Issue by hughperkins** [pytorch/pytorch#3364]
type of torch.bernoulli and torch.multinomial inconsistent:

- `torch.bernoulli` => torch.FloatTensor
- `torch.multinomial` => torch.LongTensor

I think the types should be self-consistent, ie both FloatTensor, or both LongTensor.

**Comment by zou3519** [comment]
Agreed. I'm wondering if there's some reason why they're different (they've been like this for a while it seems)

**Comment by apaszke** [comment]
So I guess the reasoning is that while Bernoulli sampling output can be always safely represented as a `float`, it's no longer true for `multinomial` if you have many classes. On the other hand, we don't have operators overloaded for cross-type ops, so returning a `LongTensor` would likely require a cast afterwards if someone wanted to use the value in the model code (which pretty much always uses `float`).

**Comment by fritzo** [comment]
In Pyro we usually treat `Categorical` variables as one-hot float vectors. It's a little easier to create a one-hot vector from a `torch.LongTensor`, since [you can](https://github.com/uber/pyro/blob/dev/pyro/distributions/categorical.py#L117)
```py
sample = torch.multinomial(probs, 1, replacement=True)
one_hot = torch.zeros_like(probs).scatter_(-1, sample, 1)
```
